upup/pkg/fi/cloudup/spotinsttasks/elastigroup.go

```diff
@@ -0,0 +1,1345 @@
+	ClusterIdentifier        *string
+}
+
+var _ fi.CompareWithID = &Elastigroup{}
+
+func (e *Elastigroup) CompareWithID() *string {
+	return e.Name
+}
+
+func (e *Elastigroup) find(svc spotinst.Service, name string) (*aws.Group, error) {
+	groups, err := svc.List(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("spotinst: failed to find elastigroup %s: %v", name, err)
+	}
+
+	var out *aws.Group
+	for _, group := range groups {
+		if group.Name() == name {
+			out = group.Obj().(*aws.Group)
+			break
+		}
+	}
+	if out == nil {
+		return nil, fmt.Errorf("spotinst: failed to find elastigroup %q", name)
+	}
+
+	return out, nil
+}
+
+var _ fi.HasCheckExisting = &Elastigroup{}
+
+func (e *Elastigroup) Find(c *fi.Context) (*Elastigroup, error) {
+	cloud := c.Cloud.(awsup.AWSCloud)
+
+	group, err := e.find(cloud.Spotinst(), *e.Name)
+	if err != nil {
+		return nil, err
+	}
+	if group == nil {
+		return nil, nil
+	}
+
+	actual := &Elastigroup{}
+	actual.ID = group.ID
+	actual.Name = group.Name
+	actual.MinSize = fi.Int64(int64(fi.IntValue(group.Capacity.Minimum)))
+	actual.MaxSize = fi.Int64(int64(fi.IntValue(group.Capacity.Maximum)))
+	actual.Orientation = group.Strategy.AvailabilityVsCost
+
+	// Compute.
+	{
+		compute := group.Compute
+		actual.Product = compute.Product
+
+		// Instance types.
+		{
+			actual.OnDemandInstanceType = compute.InstanceTypes.OnDemand
+			actual.SpotInstanceTypes = compute.InstanceTypes.Spot
+		}
+
+		// Subnets.
+		{
+			for _, zone := range compute.AvailabilityZones {
+				if zone.SubnetID != nil {
+					actual.Subnets = append(actual.Subnets, &awstasks.Subnet{ID: zone.SubnetID})
+				}
+			}
+			if subnetSlicesEqualIgnoreOrder(actual.Subnets, e.Subnets) {
+				actual.Subnets = e.Subnets
+			}
+		}
+	}
+
+	// Launch Specification.
+	{
+		lc := group.Compute.LaunchSpecification
+
+		// Image.
+		{
+			image, err := resolveImage(cloud, fi.StringValue(lc.ImageID))
+			if err != nil {
+				return nil, err
+			}
+			actual.ImageID = image.Name
+		}
+
+		// Tags.
+		{
+			if len(lc.Tags) > 0 {
+				actual.Tags = make(map[string]string)
+				for _, tag := range lc.Tags {
+					actual.Tags[*tag.Key] = *tag.Value
+				}
+			}
+		}
+
+		// Security groups.
+		{
+			for _, sg := range lc.SecurityGroupIDs {
+				actual.SecurityGroups = append(actual.SecurityGroups, &awstasks.SecurityGroup{ID: fi.String(sg)})
+			}
+		}
+
+		// Block device mappings.
+		{
+			for _, b := range lc.BlockDeviceMappings {
+				if b.EBS == nil || b.EBS.SnapshotID != nil {
+					// Not the root.
+					continue
+				}
+				actual.RootVolumeType = b.EBS.VolumeType
+				actual.RootVolumeSize = fi.Int64(int64(fi.IntValue(b.EBS.VolumeSize)))
+				actual.RootVolumeIOPS = fi.Int64(int64(fi.IntValue(b.EBS.IOPS)))
+			}
+		}
+
+		// User data.
+		{
+			if lc.UserData != nil {
+				userData, err := base64.StdEncoding.DecodeString(*lc.UserData)
+				if err != nil {
+					return nil, fmt.Errorf("spotinst: error decoding user data: %v", err)
```

Comment:
FYI: I think (hope) when this is printed it'll be printed alongside the failing task, so you likely don't need the `spotinst: ` bit.  But definitely not a blocker - our error messages repeat themselves a lot already :-)

Reply:
[Fixed](https://github.com/kubernetes/kops/pull/5922/commits/871cc42f1b35336d7e9886292e2287a3e59c5141).